src/software/thunderscope/binary_context_managers.py

```diff
@@ -0,0 +1,419 @@
+import os
+import socket
+import logging
+import psutil
+import time
+import google.protobuf.internal.encoder as encoder
+import google.protobuf.internal.decoder as decoder
+
+from proto.import_all_protos import *
+from subprocess import Popen
+from software.networking import threaded_unix_sender, networking
+from software.py_constants import *
+from extlibs.er_force_sim.src.protobuf.world_pb2 import (
+    SimulatorState,
+    SimBall,
+    SimRobot,
+)
+
+
+def is_cmd_running(command):
+    """Check if there is any running process that was launched
+    with the given command.
+
+    :param command: Command that was used to launch the process. List of strings.
+
+    """
+    for proc in psutil.process_iter():
+        try:
+            for string in command:
+                if string not in "".join(proc.cmdline()):
+                    break
+            else:
+                return True
+        except (psutil.NoSuchProcess, psutil.AccessDenied, psutil.ZombieProcess):
+            pass
+
+    return False
+
+
+class FullSystem(object):
+
+    """ Full System Binary Context Manager """
+
+    DEBUG_MODE_POLL_INTERVAL_S = 0.1
+
+    def __init__(
+        self,
+        fullsystem_runtime_dir=None,
+        debug_fullsystem=False,
+        friendly_colour_yellow=False,
+    ):
+        """Run FullSystem
+
+        :param fullsystem_runtime_dir: The directory to run the blue fullsystem in
+        :param debug_fullsystem: Whether to run the fullsystem in debug mode
+
+        """
+        self.fullsystem_runtime_dir = fullsystem_runtime_dir
+        self.debug_fullsystem = debug_fullsystem
+        self.friendly_colour_yellow = friendly_colour_yellow
+        self.fullsystem_proc = None
+
+    def __enter__(self):
+        """Enter the fullsystem context manager. 
+
+        If the debug mode is enabled then the binary is _not_ run and the
+        command to debug under gdb is printed. The  context manager will then
+        wait for the binary to be launched before continuing.
+
+        :return: fullsystem context managed instance
+
+        """
+        # Setup unix socket directory
+        try:
+            os.mkdir(self.fullsystem_runtime_dir)
+        except:
+            pass
+
+        full_system = "software/unix_full_system --runtime_dir={} {}".format(
+            self.fullsystem_runtime_dir,
+            "--friendly_colour_yellow" if self.friendly_colour_yellow else "",
+        )
+
+        if self.debug_fullsystem:
```

Comment:
I don't think this gets triggered by the `-d` flag in tbots.py

Reply:
fixed

Reply:
running debug is kinda clunky-->what if it can detect a gdb running and not start up a full system for that colour, but start up the other full system that's missing, so we can debug just one of the full systems?